### PR TITLE
Support adding submetrics to an existing compound metric

### PR DESCRIPTION
**Pull request by richard-rogers**
- Generalize compound metric `merge()` to support merging metrics with non-empty symmetric set difference of submetrics
- Dynamically add EXIF tag submetrics as they're discovered in `ImageMetric`

### CODE REVIEW

**Comment by TheMellyBee** [comment]
Just put the tag in so I can see it easier on the PR screen. Let me know when you're ready for a review.

**Review by TheMellyBee — COMMENTED** [review]
Looking good so far. I'll check back in as it becomes ready.

File: python/whylogs/extras/image_metric.py [hunk 1]
```diff
@@ -101,11 +102,32 @@ def image_based_metadata(img):
     }
 
 
+@dataclass(frozen=True)
+class ImageMetricConfig(MetricConfig):
+    allowed_exif_tags: Set[str] = field(default_factory=set)
+    forbidden_exif_tags: Set[str] = field(default_factory=set)
+
+
+def _trackable_exif_tag(exif_tag: str, allowed_tags: Set[str], forbidden_tags: Set[str]) -> bool:
```
Comment (TheMellyBee) [inline python/whylogs/extras/image_metric.py]: Doesn't this need to be tabbed in to be under the class? the ones before are tabbed in so I assume this was supposed to be apart of the class not on its own?

File: python/whylogs/extras/image_metric.py [hunk 2]
```diff
@@ -118,15 +140,28 @@ def columnar_update(self, view: PreprocessedColumn) -> OperationResult:
...
 
     @classmethod
     def zero(cls, config: MetricConfig) -> "ImageMetric":
+        if not isinstance(config, ImageMetricConfig):
+            logger.warning("ImageMetric.zero() needs an ImageMetricConfig")
+            config = ImageMetricConfig()
+
         dummy_image = new_image("HSV", (1, 1))
-        metadata = get_pil_image_metadata(dummy_image)
+        metadata = get_pil_image_metadata(dummy_image)  # any EXIF tags will be discovered as images are logged
```
Comment (TheMellyBee) [inline python/whylogs/extras/image_metric.py:163]: Wait read this again, why are you putting a dummy image through?

File: python/whylogs/extras/image_metric.py (hunk 1 above)
Comment (richard-rogers) [reply python/whylogs/extras/image_metric.py]: Stylistically I liked this better than `self._trackable_exif_tag(tag_name)`. If it starts growing a lot more arguments that are members of the `ImageMetric` I'd probably change my mind.

Comment (richard-rogers) [reply python/whylogs/extras/image_metric.py]: I might be persuaded that `self._wants_to_track(tag_name)` reads nicely...

File: python/whylogs/extras/image_metric.py (hunk 2 above)
Comment (richard-rogers) [reply python/whylogs/extras/image_metric.py:163]: It gives a nice SPOT for the image stats we track for all images (versus the data-dependent EXIF tags we discover as images are logged).

File: python/poetry.lock [hunk 3]
```diff
@@ -99,7 +99,7 @@ tests_no_zope = ["cloudpickle", "coverage[toml] (>=5.0.2)", "hypothesis", "mypy
 
 [[package]]
 name = "autoflake"
-version = "1.6.0"
+version = "1.6.1"
```
Comment (jamie256) [inline python/poetry.lock:102]: Why is a poetry.lock update in this PR? Do you need this? There are no updates to the install time dependencies in pyproject.toml

File: python/whylogs/extras/image_metric.py [hunk 4]
```diff
@@ -101,32 +102,65 @@ def image_based_metadata(img):
     }
 
 
+@dataclass(frozen=True)
+class ImageMetricConfig(MetricConfig):
+    allowed_exif_tags: Set[str] = field(default_factory=set)
+    forbidden_exif_tags: Set[str] = field(default_factory=set)
```
Comment (jamie256) [inline python/whylogs/extras/image_metric.py:108]: This capability might be interesting, but I don't know of anyone requesting this feature and it complicates the calling pattern. I thought you only had explored using ImageMetricConfig as a work-around to having no support for logging exif tags like we did in v0, and not having ability to copy metrics in a merge. Why do we need ImageMetricConfig?

File: python/whylogs/extras/image_metric.py (hunk 2 above)
Comment (jamie256) [reply python/whylogs/extras/image_metric.py:163]: If you can add to the set of submetrics as you log actual images, I don't think you need the dummy image any more, right?

File: python/whylogs/extras/image_metric.py [hunk 5]
```diff
@@ -101,32 +102,65 @@ def image_based_metadata(img):
...
         count = 0
         for image in list(chain.from_iterable(view.raw_iterator())):
             if isinstance(image, ImageType):
                 metadata = get_pil_image_metadata(image)
                 for attr in metadata.keys():
+                    if attr not in submetric_names:  # EXIF tag discovery
+                        if self._wants_to_track(attr):
+                            # TODO: add a resolver to ImageMetricConfig
+                            if isinstance(metadata[attr], int) or isinstance(metadata[attr], float):
+                                self.submetrics[attr] = DistributionMetric.zero(MetricConfig())
```
Comment (jamie256) [inline python/whylogs/extras/image_metric.py:146]: also TODO for removing MetricConfig here if its not needed (or just remove it now).

File: python/whylogs/extras/image_metric.py [hunk 6]
```diff
@@ -101,32 +102,65 @@ def image_based_metadata(img):
...
+                                self.submetrics[attr] = FrequentItemsMetric.zero(MetricConfig())
+
                     if attr in submetric_names:
                         self.submetrics[attr].columnar_update(PreprocessedColumn.apply([metadata[attr]]))
+
                 count += 1
         return OperationResult.ok(count)
 
     @classmethod
     def zero(cls, config: MetricConfig) -> "ImageMetric":
```
Comment (jamie256) [inline python/whylogs/extras/image_metric.py:157]: why is `config` required here?

File: python/whylogs/extras/image_metric.py [hunk 7]
```diff
@@ -101,32 +102,65 @@ def image_based_metadata(img):
...
+        return exif_tag in self._allowed_exif_tags
+
     def columnar_update(self, view: PreprocessedColumn) -> OperationResult:
         submetric_names = self.submetrics.keys()
         count = 0
         for image in list(chain.from_iterable(view.raw_iterator())):
             if isinstance(image, ImageType):
                 metadata = get_pil_image_metadata(image)
                 for attr in metadata.keys():
+                    if attr not in submetric_names:  # EXIF tag discovery
```
Comment (jamie256) [inline python/whylogs/extras/image_metric.py:142]: This is the key new behavior we need: discover and track features when found in images.

File: python/whylogs/extras/image_metric.py [hunk 8]
```diff
@@ -101,32 +102,65 @@ def image_based_metadata(img):
...
+                        if self._wants_to_track(attr):
+                            # TODO: add a resolver to ImageMetricConfig
```
Comment (jamie256) [inline python/whylogs/extras/image_metric.py:144]: Right, basically we have to resolve the set of metrics here similarly to how the metrics are resolved when a new column is logged.

**Review by jamie256 — CHANGES_REQUESTED** [review]
It would help to have some updates example an notebook with these kinds of changes for context and to help test.

I'm not sure we need (or I am missing context on the motivation) some parts of this PR's changes: ImageMetricConfig? poetry.lock? Can you explain or focus the changes? Maybe ImageMetricConfig can be in this PR and ignored by callers since its optional.

The change that enables users to log images and have newly discovered submetrics included without having to specify them ahead of logging is the key new thing here. This will support the logging pattern we had in v0 for images.

Mostly questions and some minor cleanup, otherwise approach looks good.

**Comment by richard-rogers** [comment]
Merged via another PR with additional work on `ImageMetric`

**Pull request closed without merging**